consensus/engine/consensus_engine.go

```diff
@@ -115,6 +115,6 @@ type Engine interface {
 		state *state.DB, txs []*types.Transaction,
 		receipts []*types.Receipt, outcxs []*types.CXReceipt,
 		incxs []*types.CXReceiptsProof, stks staking.StakingTransactions,
-		doubleSigners slash.Records,
+		doubleSigners slash.Records, sigsReady chan bool,
```

Comment:
I am not a big fan of this hack approach since there are huge risks to add multi-threading logic in the state machine which is designed to be single threaded which returns immediately. 

1. Now function `InsertChain` holds the blockchain global mutex now will probably get blocked until sigsReady timeout. Do you think it will cause other goroutines running `InsertChain` intentions to be failed unexpected. causing the locking scenario of blockchain to be extensively complicated. I would suggest not to do this "hack" coding, instead I would prefer do it in a dummy way that only finalize after collected enough commit signature.

2. Is there any code that provides a constraint on leader node cannot enter prepare phase until last block is committed?

Reply:
1. Yea, now all calls on Finalize needs to be async. That's a tradeoff of adding this feature. Any other idea to do this in a better way? Note we don't want to block/wait and do the Finalize() as a whole afterwards, because finalize contains the block reward logic for shard chains, which is expensive and could be computed earlier without the commit signatures of last beacon block.

2. We need the commit signatures to compute block reward for the new block.